x/ethbridge/keeper/msg_server.go

```diff
@@ -0,0 +1,322 @@
+package keeper
+
+import (
+	"context"
+	"strconv"
+	"sync"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/pkg/errors"
+
+	"github.com/Sifchain/sifnode/x/ethbridge/types"
+	oracletypes "github.com/Sifchain/sifnode/x/oracle/types"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the ethbridge MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = msgServer{}
+
+func (srv msgServer) Lock(goCtx context.Context, msg *types.MsgLock) (*types.MsgLockResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	logger := srv.Keeper.Logger(ctx)
+	if srv.Keeper.ExistsPeggyToken(ctx, msg.Symbol) {
+		logger.Error("pegged token can't be lock.", "tokenSymbol", msg.Symbol)
+		return nil, errors.Errorf("Pegged token %s can't be lock.", msg.Symbol)
+	}
+
+	cosmosSender, err := sdk.AccAddressFromBech32(msg.CosmosSender)
+	if err != nil {
+		return nil, err
+	}
+
+	account := srv.Keeper.accountKeeper.GetAccount(ctx, cosmosSender)
+	if account == nil {
+		logger.Error("account is nil.", "CosmosSender", msg.CosmosSender)
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, msg.CosmosSender)
+	}
+
+	if err := srv.Keeper.ProcessLock(ctx, cosmosSender, msg); err != nil {
+		logger.Error("bridge keeper failed to process lock.", errorMessageKey, err.Error())
+		return nil, err
+	}
+
+	logger.Info("sifnode emit lock event.",
+		"EthereumChainID", strconv.FormatInt(msg.EthereumChainId, 10),
+		"CosmosSender", msg.CosmosSender,
+		"CosmosSenderSequence", strconv.FormatUint(account.GetSequence(), 10),
+		"EthereumReceiver", msg.EthereumReceiver,
+		"Amount", msg.Amount.String(),
+		"Symbol", msg.Symbol,
+		"CethAmount", msg.CethAmount.String())
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.CosmosSender),
+		),
+		sdk.NewEvent(
+			types.EventTypeLock,
+			sdk.NewAttribute(types.AttributeKeyEthereumChainID, strconv.FormatInt(msg.EthereumChainId, 10)),
+			sdk.NewAttribute(types.AttributeKeyCosmosSender, msg.CosmosSender),
+			sdk.NewAttribute(types.AttributeKeyCosmosSenderSequence, strconv.FormatUint(account.GetSequence(), 10)),
+			sdk.NewAttribute(types.AttributeKeyEthereumReceiver, msg.EthereumReceiver),
```

Comment:
Same comment here if EthereumReceiver has been converted to an address type, we will need to convert it.